src/client/gemini_client.rs

```diff
@@ -0,0 +1,287 @@
+                    span: None,
+                    help: None,
+                    inner: vec![],
+                })
+            }
+        };
+
+        let mut rec: Vec<Vec<f32>> = vec![];
+        for vals in embd.embeddings {
+            rec.push(vals.values);
+        }
+
+        Ok(rec)
+    }
+
+    pub async fn ask(&self, question: String, context: Vec<String>) -> Result<String, ShellError> {
+        let url = format!(
+            "https://generativelanguage.googleapis.com/v1beta/models/gemini-pro:generateContent?key={}",
+            self.api_key
+        );
+
+        let question_with_ctx = if context.len() > 0 {
+            format!(
+                "Please answer this question: \\\"{}\\\". Using the following context: \\\"{}\\\"",
+                question,
+                context.join(" ")
+            )
+        } else {
+            question
+        };
+
+        let ask_request: AskRequest = AskRequest {
+            contents: vec![Parts {
+                parts: vec![Text {
+                    text: question_with_ctx.clone(),
+                }],
+            }],
+        };
+
+        let res = execute_request(url, ask_request).await?;
+
+        let bytes = read_response(res).await?;
+
+        let ans: AskResponse = match serde_json::from_slice(&bytes) {
+            Ok(a) => a,
+            Err(e) => {
+                return Err(ShellError::GenericError {
+                    error: format!("could not parse Gemini response: {}", e.to_string()),
+                    msg: "".to_string(),
+                    span: None,
+                    help: None,
+                    inner: vec![],
+                })
+            }
+        };
+
+        Ok(ans.candidates[0].content.parts[0].text.clone())
+    }
+}
+
+fn error_message(bytes: bytes::Bytes) -> String {
+    #[derive(Deserialize, Debug)]
+    struct ErrorResponse {
+        error: Error,
+    }
+
+    #[derive(Deserialize, Debug)]
+    struct Error {
+        message: String,
+    }
+
+    let err_msg: ErrorResponse = serde_json::from_slice(&bytes).unwrap();
+
+    err_msg.error.message
+}
+
+async fn read_response(res: Response) -> Result<Bytes, ShellError> {
+    let status = res.status().as_u16();
+    let bytes = match res.bytes().await {
+        Ok(b) => b,
+        Err(e) => {
+            return Err(ShellError::GenericError {
+                error: format!("could not read response body: {}", e.to_string()),
+                msg: "".to_string(),
+                span: None,
+                help: None,
+                inner: vec![],
+            })
+        }
+    };
+
+    if status != 200 {
+        return Err(ShellError::GenericError {
+            error: error_message(bytes),
+            msg: "".to_string(),
+            span: None,
+            help: None,
+            inner: vec![],
+        });
+    };
+
+    Ok(bytes)
+}
+
+async fn execute_request<T>(url: String, json_body: T) -> Result<Response, ShellError>
```

Comment:
Slight concerns about this not being deadlined.

Reply:
Have added a 30 second timeout

Reply:
This seems like it should be customizable or at least passed in from somewhere higher? Fine with this filed as an issue to address later though.